### PR TITLE
[FLINK-32259][table] Add built-in ARRAY_JOIN function.

### PR DESCRIPTION
**Pull request by hanyuzheng7**
## What is the purpose of the change
This is an implementation of ARRAY_JOIN

## Brief change log
Returns a string that represents the concatenation of the elements in the given array and the elements' data type in the given array is string. The delimiter is a string that separates each pair of consecutive elements of the array. The optional nullReplacement is a string that replaces null elements in the array. If nullReplacement is not specified, null elements in the array will be omitted from the resulting string. Returns null if the input array and delimiter are null.
 
- Syntax:
`array_join(array, delimiter, [, null_replacement])`

- Arguments:
array: An ARRAY to be handled.
delimiter: A STRING used to separate the concatenated array elements.
null_replacement: A STRING used to express a NULL value in the result. It is a optional

- Return:
returns a string that represents the concatenation of the elements in the given array and the elements' data type in the given array is string.
Returns null if input array or delimiter or nullReplacement are null.

- Examples:
```
> SELECT array_join(array('hello', 'world'), ' ');
hello world

> SELECT array_join(array('hello', NULL ,'world'), ' ');
hello world

> SELECT array_join(array('hello', NULL ,'world'), ' ', ',');
hello , world
```
- See also:
spark - https://spark.apache.org/docs/latest/api/sql/index.html#array_join
presto - https://prestodb.io/docs/current/functions/array.html

## Verifying this change

- This change added tests in CollectionFunctionsITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        DataType actualType = actualDataTypes.get(argumentPos);
+
+        // Check if the type is an array and its element is string
+        if (actualType.getLogicalType().getTypeRoot() == LogicalTypeRoot.ARRAY
+                && (((ArrayType) actualType.getLogicalType()).getElementType().getTypeRoot()
+                                == LogicalTypeRoot.VARCHAR
+                        || ((ArrayType) actualType.getLogicalType()).getElementType().getTypeRoot()
+                                == LogicalTypeRoot.CHAR)) {
+            return Optional.of(actualType);
+        }
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java:58]: We should rather check if we can implicitly cast to `VARCHAR`.  You can reuse: https://github.com/apache/flink/blob/fbc086f2fd63dac4af3e104ba4011e864781139a/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/StrategyUtils.java#L56

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 2]
```diff
@@ -1408,6 +1409,24 @@ public OutType arrayUnion(InType array) {
...
 
+    /**
+     * Returns a string that represents the concatenation of the elements in the given array and the
+     * elements' data type in the given array is string. The `delimiter` is a string that separates
+     * each pair of consecutive elements of the array. The optional `nullReplacement` is a string
+     * that replaces null elements in the array.
+     *
+     * <p>Returns null if input array or delimiter or nullReplacement are null.
+     */
+    public OutType arrayJoin(InType delimiter, InType... nullReplacement) {
```
[the remark this answers is not in the document]
Comment (hanyuzheng7) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java]: fixed it.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 3]
```diff
@@ -285,6 +286,20 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
...
 
+    public static final BuiltInFunctionDefinition ARRAY_JOIN =
+            BuiltInFunctionDefinition.newBuilder()
+                    .name("ARRAY_JOIN")
+                    .kind(SCALAR)
+                    .inputTypeStrategy(
+                            varyingSequence(
+                                    new ArrayOfStringArgumentTypeStrategy("ARRAY<STRING>"),
+                                    logical(LogicalTypeFamily.CHARACTER_STRING),
+                                    logical(LogicalTypeFamily.CHARACTER_STRING)))
```
[the remark this answers is not in the document]
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java]: fixed it.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java (hunk 1 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java]: fixed it

**Comment by hanyuzheng7** [comment]
@dawidwys I have already fixed all comments. You can do the second pass. Thank you.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+import static java.lang.Boolean.TRUE;
+
+/** Strategy for an argument that must be an array of strings. */
+@Internal
+public final class ArrayOfStringArgumentTypeStrategy implements ArgumentTypeStrategy {
+
+    private final String constraintMessage;
+
+    public ArrayOfStringArgumentTypeStrategy(String constraintMessage) {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java:43]: Why do we need the parameter? What other values you expect here?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(elementDataType.getLogicalType());
+    }
+
+    public @Nullable StringData eval(
+            ArrayData array, StringData delimiter, StringData... nullReplacement) {
+        try {
+            if (array == null
+                    || delimiter == null
+                    || (nullReplacement.length != 0 && nullReplacement[0] == null)) {
```
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:51]: could you add a test for that?
array != null && delimiter != null && nullReplacement[] == null

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                return null;
+            }
+            StringBuilder sb = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:54]: Ok, I will reuse BinaryStringDataUtil.concatWs function code

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java (hunk 5 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:51]:
```
.testResult(
                                call("ARRAY_JOIN", $("f0"), null),
                                "ARRAY_JOIN(f0, null)",
                                null,
                                DataTypes.STRING())
```
this one

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java [hunk 7]
```diff
@@ -479,4 +481,309 @@ private Stream<TestSetSpec> arrayUnionTestCases() {
...
+                        .testResult(
+                                call("ARRAY_JOIN", $("f0"), null),
+                                "ARRAY_JOIN(f0, null)",
+                                null,
+                                DataTypes.STRING())
+                        .testResult(
+                                call("ARRAY_JOIN", $("f0"), "+", null),
+                                "ARRAY_JOIN(f0, '+', null)",
+                                null,
+                                DataTypes.STRING())
```
Comment (hanyuzheng7) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:526]: This one is the test case of
array != null && delimiter != null && nullReplacement[] == null

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java (hunk 5 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:55]: https://github.com/apache/flink/pull/22785/files#r1244294436

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java (hunk 4 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java:43]: Ok, I will remove this argument.

**Comment by hanyuzheng7** [comment]
@dawidwys I have already fixed all comments. You can do a pass. Thank you.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java (hunk 7 above)
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:526]: ah, thanks! missed it

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        private int currentPos;
+
+        public ArrayIterator(
+                ArrayData.ElementGetter elementGetter,
+                ArrayData arrayData,
+                StringData[] nullReplacement) {
+            this.size = arrayData.size();
+            this.elementGetter = elementGetter;
+            this.arrayData = arrayData;
+            this.nullReplacement = nullReplacement;
```
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:81]: why don't we pass just the first element of an array instead? In the current approach we have unnecessary array indexing.

**Comment by dawidwys** [comment]
Current code looks mostly ok. Please rebase it and I hope we will be able to merge it in the next iteration.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java (hunk 8 above)
Comment (dawidwys) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayJoinFunction.java:81]: Still you should perform all those checks before creating the iterator. It does not change between calls to `next`.

You can do sth like:
```
if (nullReplacement.length != 0 && nullReplacement[0] != null) {
 normalizedReplacement = nullReplacement[0]  
} else {
  normalizedReplacement = null
}

new ArrayIterator(elementGetter, array, normalizedReplacement)
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    public Optional<DataType> inferArgumentType(
+            CallContext callContext, int argumentPos, boolean throwOnFailure) {
+
+        final List<DataType> actualDataTypes = callContext.getArgumentDataTypes();
+        DataType actualType = actualDataTypes.get(argumentPos);
+        if (!actualType.getLogicalType().getTypeRoot().equals(LogicalTypeRoot.ARRAY)) {
+            throw new ValidationException(
+                    "Invalid input arguments. Expected signatures are:\n"
+                            + "ARRAY_JOIN(ARRAY<STRING>, <CHARACTER_STRING>)\n"
+                            + "ARRAY_JOIN(ARRAY<STRING>, <CHARACTER_STRING>, <CHARACTER_STRING>)");
```
[the remark this answers is not in the document]
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayOfStringArgumentTypeStrategy.java:49]: fixed it.

**Comment by hanyuzheng7** [comment]
@dawidwys ok, wait for https://github.com/apache/flink/pull/22730 merged and then rebase and resolve conflicts.
Because array_max change the code at same position as array_join.

**Comment by hanyuzheng7** [comment]
@flinkbot run Azure

**Comment by hanyuzheng7** [comment]
@dawidwys I have already rebased it and solved conflicts and run CI test. It's ready to merge now. Thank you!

**Review by dawidwys — APPROVED** [review]
LGTM

**Comment by hanyuzheng7** [comment]
> What should be the result for joining two empty arrays?

why we want join two array?

**Comment by hanyuzheng7** [comment]
> What should be the result for joining two empty arrays?

we don't need join two empty arrays, 

for empty array
```
.testSqlValidationError(
                                "ARRAY_JOIN()",
                                "No match found for function signature ARRAY_JOIN()")
```

**Comment by dawidwys** [comment]
Sorry, I meant an empty array 🤦

**Comment by hanyuzheng7** [comment]
> Sorry, I meant an empty array 🤦

NP, thank you for reviewing my code.

**Pull request merged**